gnome-initial-setup/demo/eos-demo-video.c

```diff
@@ -0,0 +1,212 @@
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *   Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include "config.h"
+
+#include <eos-demo-video.h>
+
+typedef struct _GisDemoVideoPrivate
+{
+  GtkWindow *main_window;
+  GstElement *element;
+  GstElement *player;
+  GtkWidget *widget;
+  GstBus *bus;
+} GisDemoVideoPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (GisDemoVideo, gis_demo_video, GTK_TYPE_APPLICATION)
+
+static gboolean
+on_demo_video_event (GtkWidget *widget,
+                     GdkEvent  *event,
+                     gpointer  user_data)
+{
+  if (event->type == GDK_BUTTON_PRESS ||
+      event->type == GDK_2BUTTON_PRESS ||
+      event->type == GDK_3BUTTON_PRESS ||
+      event->type == GDK_KEY_PRESS ||
+      event->type == GDK_MOTION_NOTIFY)
+    {
+      gtk_widget_destroy (widget);
+    }
+
+  return FALSE;
+}
+
+static gboolean
+on_video_bus_event (GstBus     *bus,
+                    GstMessage *msg,
+                    gpointer   user_data)
+{
+  GisDemoVideo *demo_video = GIS_DEMO_VIDEO (user_data);
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+
+  switch (GST_MESSAGE_TYPE (msg))
+    {
+    case GST_MESSAGE_EOS:
+      gst_element_set_state (priv->player, GST_STATE_PAUSED);
+      gst_element_seek_simple (priv->player,
+                               GST_FORMAT_TIME,
+                               GST_SEEK_FLAG_FLUSH |
+                               GST_SEEK_FLAG_ACCURATE,
+                               0);
+      gst_element_set_state (priv->player, GST_STATE_PLAYING);
+    default:
+      break;
+    }
+
+  return G_SOURCE_CONTINUE;
+}
+
+#define DEMO_VIDEO_FILE_URI "file://" DEMOVIDEODATADIR "/eos-demo-video.ogv"
+
+static void
+create_demo_window (GisDemoVideo *demo_video)
+{
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+  GtkApplication *application = GTK_APPLICATION (demo_video);
+  GtkWidget *window = NULL;
+
+  priv->main_window = GTK_WINDOW (gtk_application_window_new (application));
+  gtk_widget_set_events (GTK_WIDGET (priv->main_window),
+                         GDK_POINTER_MOTION_MASK |
+                         GDK_KEY_PRESS_MASK |
+                         GDK_BUTTON_PRESS_MASK); 
+  gtk_window_fullscreen (priv->main_window);
+  gtk_application_add_window (application,
+                              priv->main_window);
+
+  /* Translators: this text appears when the user invokes
+   * some action that is blocked because the demo video
+   * is showing */
+  gtk_application_inhibit (application,
+                           priv->main_window,
+                           GTK_APPLICATION_INHIBIT_LOGOUT |
+                           GTK_APPLICATION_INHIBIT_SWITCH |
+                           GTK_APPLICATION_INHIBIT_IDLE,
+                           _("Demonstration video"));
+
+  gtk_container_add (GTK_CONTAINER (priv->main_window), priv->widget);
+
+  gtk_widget_show (priv->widget);
```

Comment:
Of course, must have been a leftover from then we were still handling reparenting.